app/src/main/java/com/lightricks/feedexercise/data/remote/RemoteFeedDataSource.kt

```diff
@@ -0,0 +1,37 @@
+package com.lightricks.feedexercise.data.remote
+
+import com.lightricks.feedexercise.data.toFeed
+import com.lightricks.feedexercise.domain.FeedError
+import com.lightricks.feedexercise.domain.FeedSuccess
+import com.lightricks.feedexercise.network.FeedApiService
+import com.lightricks.feedexercise.network.UrlProvider
+import com.lightricks.feedexercise.util.result.toFailureResult
+import com.lightricks.feedexercise.util.result.toSuccessResult
+import java.net.UnknownHostException
+import com.lightricks.feedexercise.util.result.Result
+
+interface RemoteFeedDataSource {
+    suspend fun getFeed(): Result<FeedError, FeedSuccess>
+}
+
+internal class RetrofitFeedDataSource(
+    private val apiService: FeedApiService,
+    private val thumbnailBaseUrlProvider: UrlProvider
+) : RemoteFeedDataSource {
+    override suspend fun getFeed(): Result<FeedError, FeedSuccess> {
+        return try {
+            val response = apiService.getFeed()
+            return when {
+                response.isSuccessful && response.body() != null -> requireNotNull(response.body())
+                    .toFeed(thumbnailBaseUrlProvider)
+                    .let { feed -> FeedSuccess(feed = feed) }
+                    .toSuccessResult()
+                else -> FeedError.NetworkError.toFailureResult()
+            }
+        } catch (e: UnknownHostException) {
+            FeedError.NetworkError.toFailureResult()
+        } catch (e: Throwable) {
+            FeedError.UnknownError.toFailureResult()
+        }
```

Comment:
following https://github.com/pasha-dektev/feed-exercise/pull/1#discussion_r895160855
The bad thing when we work with custom class instead of exception is that we lost the stack trace. Here you do nothing with the exception, so if we want to debug the cause it will be very hard to do it without the stack trace.

Reply:
Are you suggesting to throw the exception so that the app will crash? If not - than if we catch the exception than we wont be able to debug it anyway.

Reply:
Crash? That depends on the caller of how it wants to deal with this exception. What I am saying is that we can rethrow a custom exception of this specific domain (by wrapping the caught one) and the caller will decide how to deal with this exception. Logging it and/or showing a popup/message to the user... That way we not lost data (stack trace).

Reply:
You can also return the exception wrapped by the kotlin Result (I don't like it) because we can fast not to do anything with the exception it contains.

Reply:
> Are you suggesting to throw the exception so that the app will crash? If not - than if we catch the exception than we wont be able to debug it anyway.

You can debug one you have the stack trace...

Reply:
I understand what you are saying. This approach is good also. But I prefer the way I did it. 
1. It gives you freedom to wrap/not wrap the exception if you really need it and handle it somewhere else.
2. Handling the result is much more readable because you just have to iterate through sealed class - no try catch and throws
3. You dont have to remember to handle exceptions. Sealed class wont let you forget that there is a specific custom error